### PR TITLE
Feature display the winning path

### PR DESCRIPTION
**Pull request by atilbian**
- [x] Extract the winning path from a given board
- [x] Add the winning path in the game state struct
- [x] Display the winning path in ui at the end of the game

Related to : https://trello.com/c/Eoq4ax1k/13-1-as-norbert-when-i-play-a-game-and-i-win-i-want-the-app-to-show-the-winning-path

## Demo
demo-feature-wining-path

### CODE REVIEW

**Comment by fzaninotto** [comment]
There is a visual "jank" when the game ends. Could you fix it?

File: src/handlebars/helpers.ts [hunk 1]
```diff
@@ -14,9 +14,25 @@ export function registerHandlebarsHelpers() {
...
+  handlebars.registerHelper(
+    'isCellInWinningPath',
+    function (
+      winningPath: Coordinates[],
+      cellXCoord: number,
+      cellYCoord: number,
+    ) {
+      return (
+        winningPath &&
+        winningPath.filter(
```
Comment (fzaninotto) [inline src/handlebars/helpers.ts]: do you know ythe `array.some()` function?

File: views/game.hbs [hunk 2]
```diff
@@ -35,7 +43,8 @@
           </label>
         </form>
         {{else}}
-        <div class="cell" data-color="{{this.value}}">
+        <div class="cell" data-color="{{this.value}}"
+          data-winningCell="{{isCellInWinningPath @root.state.winningPath @index @../index}}">
```
Comment (fzaninotto) [inline views/game.hbs:45]: I find the syntas `@../index` cryptic. Can't you name the index variable in the `#each` block?

**Review by fzaninotto — CHANGES_REQUESTED** [review]

File: src/common/pathfinding.ts [hunk 3]
```diff
@@ -1,14 +1,28 @@
...
   endNodeId: string,
-): boolean {
+): { hasWon: boolean; winningPath: Coordinates[] } {
   const dijkstra = new jkstra.algos.Dijkstra(hexBoardGraph.graph);
-  return !!dijkstra.shortestPath(
+  const shortestPath = dijkstra.shortestPath(
     hexBoardGraph.vertexMap.get(startNodeId),
     hexBoardGraph.vertexMap.get(endNodeId),
   );
+  if (!!shortestPath) {
```
Comment (slax57) [inline src/common/pathfinding.ts]: `!!` is useful for ensuring you return a boolean, but here I think `if (shortestPath)` is enough

File: src/handlebars/helpers.ts [hunk 4]
```diff
@@ -14,9 +14,25 @@ export function registerHandlebarsHelpers() {
...
+          (cell) => cell.x == cellXCoord && cell.y == cellYCoord,
+        ).length == 1
```
Comment (slax57) [inline src/handlebars/helpers.ts]:
```suggestion
        winningPath.contains(
          { x: cellXCoord , y: cellYCoord }
        )
```

wouldn't that work?

File: src/handlebars/helpers.ts (hunk 1 above)
Comment (atilbian) [reply src/handlebars/helpers.ts]: Nop!

File: views/game.hbs (hunk 2 above)
Comment (slax57) [inline views/game.hbs:45]: Oh I didn't know about `@root`, that's great! 😃

Comment (atilbian) [reply views/game.hbs:45]: Unfortunately I don't find a way to do it.

File: src/handlebars/helpers.ts (hunk 4 above)
Comment (atilbian) [reply src/handlebars/helpers.ts]: I don't think contains exist but François gave another working solution :https://github.com/marmelab/Hex/pull/26#discussion_r809888375

**Pull request merged**